### PR TITLE
[SYCL] Make host task blocking and detach empty command. Part 1

### PR DESCRIPTION
**Pull request by KseniyaTikhomirova**
It is intended to solve existing gap in implementation that kernel enqueue depending on host task via depends_on will wait for host task completion in its own enqueue. I want to change this approach and make it similar with the one we use when we have memory dependencies.
It also should help with cleanup process because now host task is represented with two instances - ExecCGCommand and EmptyCommand that introduces leaks in some scenarios.

Signed-off-by: Tikhomirova, Kseniya <kseniya.tikhomirova@intel.com>

### CODE REVIEW

**Comment by hdelan** [comment]
Could you give a brief description of what this PR is achieving and why it is needed? Thanks

**Comment by KseniyaTikhomirova** [comment]
> Could you give a brief description of what this PR is achieving and why it is needed? Thanks

It is intended to solve existing gap in implementation that kernel enqueue depending on host task via depends_on will wait for host task completion in its own enqueue. I want to change this approach and make it similar with the one we use when we have memory dependencies.
It also should help with cleanup process because now host task is represented with two instances - ExecCGCommand and EmptyCommand that introduces leaks in some scenarios.

TO conclude this PR (not fully ready yet) has intension to:
1) make command enqueue to not wait for host task completion in its enqueueImp and to be enqueued on host task completion
2) fix some memory leaks related to pair host task + empty task
3) ideally to remove empty task at all as class

Please let me know if you need more details or have any concerns.

**Comment by KseniyaTikhomirova** [comment]
fixes https://github.com/intel/llvm/issues/7146

File: sycl/source/detail/scheduler/commands.hpp [hunk 1]
```diff
@@ -142,8 +142,20 @@ class Command {
     return MEnqueueStatus == EnqueueResultT::SyclEnqueueSuccess;
   }
 
+  // Shows that command could not be enqueued, now it may be true for empty task
+  // only
   bool isEnqueueBlocked() const {
-    return MEnqueueStatus == EnqueueResultT::SyclEnqueueBlocked;
+    return MIsBlockable && MEnqueueStatus == EnqueueResultT::SyclEnqueueBlocked;
+  }
+  // Shows thst command could be enqueud, but is blocking enqueue of all
```
[the remark this answers is not in the document]
Comment (KseniyaTikhomirova) [reply sycl/source/detail/scheduler/commands.hpp]: fixed in https://github.com/intel/llvm/pull/6995/commits/e213e81758f368d7b9903596df404a06b99bdc08

File: sycl/source/detail/scheduler/commands.hpp [hunk 2]
```diff
@@ -325,6 +337,14 @@ class Command {
   /// Indicates that the node will be freed by cleanup after enqueue. Such nodes
   /// should be ignored by other cleanup mechanisms.
   bool MPostEnqueueCleanup = false;
+
+  /// Contains list of commands that depends on the host command explicitly (by
+  /// depends_on). Not involved into cleanup process since it is one-way link
+  /// and not holds resources.
```
[the remark this answers is not in the document]
Comment (KseniyaTikhomirova) [reply sycl/source/detail/scheduler/commands.hpp]: fixed in https://github.com/intel/llvm/pull/6995/commits/e213e81758f368d7b9903596df404a06b99bdc08

File: sycl/source/detail/scheduler/scheduler.hpp [hunk 3]
```diff
@@ -786,7 +790,11 @@ class Scheduler {
     /// the lock is left in locked state.
     static bool enqueueCommand(Command *Cmd, EnqueueResultT &EnqueueResult,
                                std::vector<Command *> &ToCleanUp,
+                               Command *RootCommand,
                                BlockingT Blocking = NON_BLOCKING);
+
+    static bool handleBlockingCmd(Command *Cmd, EnqueueResultT &EnqueueResult,
```
[the remark this answers is not in the document]
Comment (KseniyaTikhomirova) [reply sycl/source/detail/scheduler/scheduler.hpp:809]: fixed in https://github.com/intel/llvm/pull/6995/commits/e213e81758f368d7b9903596df404a06b99bdc08

File: sycl/source/detail/scheduler/scheduler.hpp [hunk 4]
```diff
@@ -786,7 +790,11 @@ class Scheduler {
     /// the lock is left in locked state.
     static bool enqueueCommand(Command *Cmd, EnqueueResultT &EnqueueResult,
                                std::vector<Command *> &ToCleanUp,
+                               Command *RootCommand,
```
[the remark this answers is not in the document]
Comment (KseniyaTikhomirova) [reply sycl/source/detail/scheduler/scheduler.hpp:794]: fixed in https://github.com/intel/llvm/pull/6995/commits/e213e81758f368d7b9903596df404a06b99bdc08

File: sycl/unittests/scheduler/EnqueueWithDependsOnDeps.cpp [hunk 5]
```diff
@@ -63,155 +37,211 @@ bool CheckTestExecutionRequirements(const platform &plt) {
...
-    "SYCL_DISABLE_POST_ENQUEUE_CLEANUP";
+enum TestCGType { KERNEL_TASK = 0x00, HOST_TASK = 0x01 };
 
-TEST_F(SchedulerTest, EnqueueNoMemObjTwoHostTasks) {
-  // Checks enqueue of two dependent host tasks
+class DependsOnTests : public ::testing::Test {
+protected:
+  void SetUp() {
+    platform Plt = Mock.getPlatform();
+    if (!CheckTestExecutionRequirements(Plt))
```
[the remark this answers is not in the document]
Comment (KseniyaTikhomirova) [reply sycl/unittests/scheduler/EnqueueWithDependsOnDeps.cpp:46]: It is already using Mock plugin now. Look at the line 45. CheckTestExecutionRequirements mostly checks that we have not only host. Mock Plugin is opencl by default:
PiMock.hpp:197: PiMock(backend Backend = backend::opencl)

**Comment by KseniyaTikhomirova** [comment]
@intel/llvm-gatekeepers

**Comment by pvchupin** [comment]
@KseniyaTikhomirova, post-commit issue: https://github.com/intel/llvm/actions/runs/3689466995/jobs/6245440471
@aelovikov-intel is looking into it.

**Comment by aelovikov-intel** [comment]
https://github.com/intel/llvm/pull/7770, hopefully.

**Pull request merged**